cou/apps/auxiliary.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2023 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Auxiliary application class."""
+import logging
+from typing import Optional
+
+from cou.apps.app import AppFactory, OpenStackApplication
+from cou.exceptions import ApplicationError, HaltUpgradePlanGeneration
+from cou.steps import UpgradeStep
+from cou.utils.juju_utils import async_upgrade_charm
+from cou.utils.openstack import LTS_SERIES, OPENSTACK_TO_TRACK_MAPPING, OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@AppFactory.register_application(["rabbitmq-server", "vault", "mysql", "ovn"])
+class OpenStackAuxiliaryApplication(OpenStackApplication):
+    """Application for charms that can have multiple OpenStack releases for a workload."""
+
+    def openstack_to_track(self, os_release: OpenStackRelease) -> str:
+        """Find the track of auxiliary charms by Ubuntu release and OpenStack release codename.
+
+        :param os_release: OpenStack release to track.
+        :type os_release: OpenStackRelease
+        :raises ApplicationError: When there is no track compatible.
+        :return: The track of the auxiliary charm.
+        :rtype: str
+        """
+        try:
+            return OPENSTACK_TO_TRACK_MAPPING[self.series][self.family][os_release.codename]
+        except KeyError as exc:
+            raise ApplicationError(
+                f"Cannot find a track of '{self.charm}' for {os_release.codename}"
+            ) from exc
+
+    def os_origin_config(self, target: OpenStackRelease) -> Optional[OpenStackRelease]:
```

Comment:
it took me a bit to understand this function. You want to understand which OpenStack release is contained in the APT sources configured in this charm. Could we rename it to something more obvious?
Additionally, shouldn't this also be part of cou.utils.openstack? The resolution algorithm does not change among applications.

Reply:
What do you think about renaming to `os_origin_from_apt_sources`? 
It's indeed doesn't change among apps. I think it's better if we move to `OpenStackApplication` instead of `cou.utils.openstack`.

Reply:
yeah leaving it in `OpenStackApplication` feels more right after all. You chosen name is better, but perhaps we can further refine it. How about `apt_source_codename` and making it a property (if you drop the target parameter)?

Reply:
Done